### PR TITLE
enable serving metrics scraping

### PR DESCRIPTION
**Pull request by deepio-oc**
Screen Shot 2022-01-31 at 11 12 44 AM

### CODE REVIEW

File: kustomize/monitoring/base/monitoring.yaml [hunk 1]
```diff
@@ -125,6 +125,93 @@ data:
...
         target_label: kubernetes_name
+    - job_name: kfserving-monitoring
+      honor_timestamps: true
+      scrape_interval: 15s
+      scrape_timeout: 10s
+      metrics_path: /metrics
+      scheme: http
+      follow_redirects: true
+      relabel_configs:
+      - source_labels: [__meta_kubernetes_service_label_networking_internal_knative_dev_serviceType]
```
Comment (songole) [inline kustomize/monitoring/base/monitoring.yaml:142]: Are these labels on inference service deployments?

File: kustomize/monitoring/base/monitoring.yaml [hunk 2]
```diff
@@ -125,6 +125,93 @@ data:
...
+        separator: ;
+        regex: Private
+        replacement: $1
```
Comment (songole) [inline kustomize/monitoring/base/monitoring.yaml:145]: What would be $1?

File: kustomize/monitoring/base/monitoring.yaml [hunk 3]
```diff
@@ -125,6 +125,93 @@ data:
...
+      - source_labels: [__meta_kubernetes_endpoint_port_name]
+        separator: ;
+        regex: http-usermetric
+        replacement: $1
+        action: keep
+      - source_labels: [__meta_kubernetes_endpoint_address_target_kind, __meta_kubernetes_endpoint_address_target_name]
+        separator: ;
+        regex: Node;(.*)
+        target_label: node
+        replacement: ${1}
```
Comment (songole) [inline kustomize/monitoring/base/monitoring.yaml:156]: What is $1 & ${1} ?

Don't we create our own labels?

File: kustomize/monitoring/base/monitoring.yaml (hunk 1 above)
Comment (deepio-oc) [reply kustomize/monitoring/base/monitoring.yaml:142]: this configuration is taken from following Prometheus Setup from https://github.com/kserve/kserve/tree/master/docs/samples/metrics-and-monitoring

File: kustomize/monitoring/base/monitoring.yaml (hunk 3 above)
Comment (deepio-oc) [reply kustomize/monitoring/base/monitoring.yaml:156]: I have added deployment name and the component name. We can add more if needed

File: kustomize/monitoring/base/monitoring.yaml [hunk 4]
```diff
@@ -370,7 +469,27 @@ spec:
     app: dkube-exporter-deploy
   sessionAffinity: None
   type: ClusterIP
-
+---
```
Comment (deepio-oc) [inline kustomize/monitoring/base/monitoring.yaml:472]: add a comment

**Comment by songole** [comment]
A few comments based on the review 
- Need to honor the polling interval configured 
- Moving Alert manager configuration to controller. We delete rules when a monitor is stopped, create rules when started, update when rules are created/updated when monitor is in running state. This can be done in a separate go routine
- Soft thresholds should be recorded as 'at risk'
- Every time a rule is breached, can we record that in prometheus, so we can show when the rules were breached and when an email was sent. The rule should be read as alert with a follow up action. Follow up action is something like  'if rule fires 3 times successively, send an email'

**Comment by deepio-oc** [comment]
OSM, what does "- Need to honor the polling interval configured " means ?

Other comments doesn't apply to this PR (will need to be implemented in gpuaas repo). Could we merge this so that it is part of next build ?

**Comment by songole** [comment]
Polling interval as configured in the health monitor configuration.

3rd & 4th are you doing in the next PR?
Can you add more comments in the next PR to the code?

**Comment by songole** [comment]
We can do the polling interval in the next PR.

**Pull request merged**